### ISSUE TITLE
No errors on typed function internals if function definition itself is untyped; I'd argue that's probably incorrect behavior, but at the very least it's *super* confusing and *very* hard to diagnose when it happens

### ISSUE THREAD

**Issue by posita** [python/mypy#10895]
### Issue description

This doesn't generate an error:

```python
# test_case7.py
from typing import List

def untyped_func_def_with_typed_body():
    a: List[int] = [0, 1, 2]
    key: str = "asdf"
    a[key]  # <- No error, despite internals being typed
```

This does:

```python
# test_case8.py
from typing import List

def typed_func_def_with_typed_body() -> None:
    a: List[int] = [0, 1, 2]
    key: str = "asdf"
    a[key]  # <- *Now* this generates an error
```

That maybe _**technically**_  "[documented](https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code)", but that's _**really**_ esoteric and completely unintuitive. I've spent _**hours**_ chasing this down.

---

### Archived original flailing to arrive at core problem

_UPDATE 2: Okay, this just keeps getting weirder and weirder. Consider:_

```python
# test_case3.py

def func():
    a = [0, 1, 2]
    a["asdf"]  # <- This doesn't generate an error

b = [0, 1, 2]
b["asdf"]  # <- But this does?!
```

Runtime:

```
% mypy --config=/dev/null test_case3.py
/dev/null: No [mypy] section in config file
test_case3.py:8: error: No overload variant of "__getitem__" of "list" matches argument type "str"
test_case3.py:8: note: Possible overload variants:
test_case3.py:8: note:     def __getitem__(self, SupportsIndex) -> int
test_case3.py:8: note:     def __getitem__(self, slice) -> List[int]
Found 1 error in 1 file (checked 1 source file)
```

_UPDATE 1: You don't need the `@overload`s. It_ **still** _doesn't work without them. Here's the simplified case:_

```python
# test_case2.py
from typing import List, Union

class GetItemTest:
    def __getitem__(self, key: Union[int, slice]) -> Union[int, List[int]]:
        return list(range(10))[key]

def test_a():
    a = GetItemTest()
    a["asdf"]  # <- Why isn't this a typing error?
```

Original case (same effect):

```python
# test_case1.py
from typing import List, Union, overload

class GetItemTest:
    @overload
    def __getitem__(self, key: int) -> int:
        ...
    @overload
    def __getitem__(self, key: slice) -> List[int]:
        ...
    def __getitem__(self, key: Union[int, slice]) -> Union[int, List[int]]:
        return list(range(10))[key]

def test_a():
    a = GetItemTest()
    a["asdf"]  # <- Why isn't this a typing error?
```

Runtime:

```
% python --version
Python 3.9.6
% pip list
Package           Version
----------------- --------
mypy              0.910
mypy-extensions   0.4.3
pip               21.2.1
setuptools        57.4.0
toml              0.10.2
typing-extensions 3.10.0.0
% mypy --version
mypy 0.910
% mypy --config=/dev/null test_case1.py
/dev/null: No [mypy] section in config file
Success: no issues found in 1 source file
```

I can't prove it, but I **_feel_** like this used to work. It seems pretty basic.

**Comment by posita** [comment]
With `--verbose` if helpful:

```
% mypy --verbose --config=/dev/null test_case.py
/dev/null: No [mypy] section in config file

LOG:  Mypy Version:           0.910
LOG:  Config File:            /dev/null
LOG:  Configured Executable:  /.../bin/python
LOG:  Current Executable:     /.../bin/python
LOG:  Cache Dir:              .mypy_cache
LOG:  Compiled:               True
LOG:  Exclude:                
LOG:  Found source:           BuildSource(path='test_case.py', module='test_case', has_text=False, base_dir='/...')
LOG:  Metadata fresh for test_case: file test_case.py
LOG:  Metadata fresh for typing: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/typing.pyi
LOG:  Metadata fresh for builtins: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/builtins.pyi
LOG:  Metadata fresh for collections: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/collections/__init__.pyi
LOG:  Metadata fresh for sys: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/sys.pyi
LOG:  Metadata fresh for abc: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/abc.pyi
LOG:  Metadata fresh for types: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/types.pyi
LOG:  Metadata fresh for _typeshed: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/_typeshed/__init__.pyi
LOG:  Metadata fresh for ast: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/ast.pyi
LOG:  Metadata fresh for io: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/io.pyi
LOG:  Metadata fresh for typing_extensions: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/typing_extensions.pyi
LOG:  Metadata fresh for _collections_abc: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/_collections_abc.pyi
LOG:  Metadata fresh for importlib.abc: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/importlib/abc.pyi
LOG:  Metadata fresh for importlib.machinery: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/importlib/machinery.pyi
LOG:  Metadata fresh for array: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/array.pyi
LOG:  Metadata fresh for mmap: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/mmap.pyi
LOG:  Metadata fresh for os: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/os/__init__.pyi
LOG:  Metadata fresh for _ast: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/_ast.pyi
LOG:  Metadata fresh for codecs: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/codecs.pyi
LOG:  Metadata fresh for importlib: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/importlib/__init__.pyi
LOG:  Metadata fresh for posix: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/posix.pyi
LOG:  Metadata fresh for subprocess: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/subprocess.pyi
LOG:  Metadata fresh for os.path: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/os/path.pyi
LOG:  Metadata fresh for posixpath: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/posixpath.pyi
LOG:  Metadata fresh for genericpath: file /.../lib/python3.9/site-packages/mypy/typeshed/stdlib/genericpath.pyi
LOG:  Loaded graph with 25 nodes (0.017 sec)
LOG:  Found 2 SCCs; largest has 24 nodes
LOG:  2 fresh SCCs (25 nodes) left in queue (and will remain unprocessed)
LOG:  Build finished in 0.023 seconds with 0 modules, and 0 errors
Success: no issues found in 1 source file
```

**Comment by posita** [comment]
I have now verified this happens with Python versions 3.7.11, 3.8.11, and 3.9.6.

**Comment by pranavrajpal** [comment]
I think this is because of https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code. Does running mypy with `--check-untyped-defs` fix this?

**Comment by posita** [comment]
> I think this is because of https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code. Does running mypy with `--check-untyped-defs` fix this?

`--check-untyped-defs` "fixes" this (in a sense):

```python
% rm -fr .mypy_cache ; mypy --config=/dev/null --check-untyped-defs test_case3.py
/dev/null: No [mypy] section in config file
test_case3.py:5: error: No overload variant of "__getitem__" of "list" matches argument type "str"
test_case3.py:5: note: Possible overload variants:
test_case3.py:5: note:     def __getitem__(self, SupportsIndex) -> int
test_case3.py:5: note:     def __getitem__(self, slice) -> List[int]
test_case3.py:8: error: No overload variant of "__getitem__" of "list" matches argument type "str"
test_case3.py:8: note: Possible overload variants:
test_case3.py:8: note:     def __getitem__(self, SupportsIndex) -> int
test_case3.py:8: note:     def __getitem__(self, slice) -> List[int]
Found 2 errors in 1 file (checked 1 source file)
```

I still think this is a bug, for two reasons. First, the fact that it works at the module level and not the function level is just _weird_. Second, I **_swear_** this used to work with classes that defined their own `__getitem__` methods. (See my `test_case2.py` example in the original report.) Also, this still doesn't work without `--check-untyped-defs`:

```python
# test_case4.py
from typing import List

def func():
    a: List[int] = [0, 1, 2]
    key: str = "asdf"
    a[key]  # <- This doesn't generate an error

func()
b: List[int] = [0, 1, 2]
key: str = "asdf"
b[key]  # <- But this does?!
```

Runtime:

```
% rm -fr .mypy_cache ; mypy --config=/dev/null test_case4.py
/dev/null: No [mypy] section in config file
test_case4.py:12: error: No overload variant of "__getitem__" of "list" matches argument type "str"
test_case4.py:12: note: Possible overload variants:
test_case4.py:12: note:     def __getitem__(self, SupportsIndex) -> int
test_case4.py:12: note:     def __getitem__(self, slice) -> List[int]
Found 1 error in 1 file (checked 1 source file)
```

**Comment by posita** [comment]
Ugh, talk about a super sharp edge. This works as expected (since I typed the enclosing function):

```python
# test_case5.py
from typing import List

def func() -> None:
    a: List[int] = [0, 1, 2]
    key: str = "asdf"
    a[key]  # <- *Now* this generates an error

func()
b: List[int] = [0, 1, 2]
key: str = "asdf"
b[key]  # <- So does this
```

**Comment by posita** [comment]
As does this:

```python
# test_case6.py
from typing import List, Union

class GetItemTest:
    def __getitem__(self, key: Union[int, slice]) -> Union[int, List[int]]:
        return list(range(10))[key]

def test_a() -> None:
    a = GetItemTest()
    a["asdf"]  # <- errors
```

**Comment by erictraut** [comment]
This is yet another example that supports my thinking that `--check-untyped-defs` should default to on. See discussion [here](https://github.com/python/mypy/issues/3948#issuecomment-889946321).

**Comment by posita** [comment]
> This is yet another example that supports my thinking that `--check-untyped-defs` should default to on. See discussion [here](https://github.com/python/mypy/issues/3948#issuecomment-889946321).

Maybe as a work-around, but functions seem like pretty arbitrary boundaries to try and walk this particular line, especially where bodies contain explicit type annotations. I can't see how ignoring those can be considered correct under any setting. In other words, yeah, totally make the sane thing the default thing, but I'm pretty sure this is still a bug.

**Comment by JelleZijlstra** [comment]
It's documented that mypy doesn't type check unannotated functions (https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code). I don't disagree with you that this is unintuitive, but the other issue is a sufficient place to discuss that.

**Comment by posita** [comment]
> It's documented that mypy doesn't type check unannotated functions (https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code). I don't disagree with you that this is unintuitive, but the other issue is a sufficient place to discuss that.

I think mine is a separate issue, namely that typed function bodies shouldn't qualify as "untyped". I don't think the documentation is even accurate. Consider:

```python
# test_case9.py
from typing import Callable
a: Callable[[], int]

def _untyped_init_a():
    global a
    a = lambda: [0, 1, 2]["key"]  # errors

def _untyped_init_b():
    global b
    b = lambda: [0, 1, 2]["key"]  # doesn't error
```

According to the docs, `_untyped_init_a` shouldn't error, but it does. Un-annotated function bodies are treated inconsistently where `untyped_defs` is `False`.
